### PR TITLE
8336741: Optimize LocalTime.toString with StringBuilder.repeat

### PR DESCRIPTION
**Pull request by wenshao**
```java
class LocalTime {
    public String toString() {
        // ...
                if (nanoValue % 1000_000 == 0) {
                    buf.append(Integer.toString((nanoValue / 1000_000) + 1000).substring(1));
                } else if (nanoValue % 1000 == 0) {
                    buf.append(Integer.toString((nanoValue / 1000) + 1000_000).substring(1));
                } else {
                    buf.append(Integer.toString((nanoValue) + 1000_000_000).substring(1));
                }
       // ...
    }
}
```
Currently, LocalTime.toString handles nanos by adding a value and then subString(1) to fill it with zeros. Using StringBuilder.repeat is more concise and has better performance.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8336741](https://bugs.openjdk.org/browse/JDK-8336741): Optimize LocalTime.toString with StringBuilder.repeat (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20232/head:pull/20232` \
`$ git checkout pull/20232`

Update a local copy of the PR: \
`$ git checkout pull/20232` \
`$ git pull https://git.openjdk.org/jdk.git pull/20232/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20232`

View PR using the GUI difftool: \
`$ git pr show -t 20232`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20232.diff">https://git.openjdk.org/jdk/pull/20232.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20232#issuecomment-2236290235)

### CODE REVIEW

**Comment by wenshao** [comment]
printFixedWidth cannot meet the needs of append nanos, For example, if the value is 1000, the output is `.000001`, not `.000001000`

**Comment by liach** [comment]
This particular usage can become:
```java
buf.append('.');
if (nanoValue % 1000_000 == 0) {
    printFixedWidth(buf, 3, nanoValue / 1000_000);
} else if (nanoValue % 1000 == 0) {
    printFixedWidth(buf, 6, nanoValue / 1000);
} else {
    printFixedWidth(buf, 9, nanoValue);
}
```

**Comment by RogerRiggs** [comment]
The "better performance" claim is unsubstantiated.  By how much.
The change introduces non-local dependencies that make the code harder to follow. (Shared secrets).
In a typical application, the improvement would not be noticeable and makes the code harder to maintain.

**Comment by wenshao** [comment]
The current logic of adding a value and then substring(1) is not clear, and this approach is not elegant at all. I think the logic of changing to repeat('0') is concise and clear.

Using repeat implementation will reduce two object allocations, Integer.toString & subString(1), and the time consumption will be 1/3 of the original.

If JLA is not recommended, we can add a stringSize method in LocalTime, similar to DateTimeFormatterBuilder.

**Comment by liach** [comment]
It seems that `DecimalDigit.size` was accidentally removed in 03e84178ebfd2ca48b89d65d8f3c291e0c622fb5. You are welcome to add it back and drop the dependency on `JavaLangAccess`.

File: src/java.base/share/classes/java/time/LocalTime.java [hunk 1]
```diff
@@ -1640,13 +1644,19 @@ public String toString() {
...
             if (nanoValue > 0) {
                 buf.append('.');
+                int zeros = 9 - JLA.stringSize(nanoValue);
+                if (zeros > 0) {
+                    buf.repeat('0', zeros);
+                }
+                int digits;
                 if (nanoValue % 1000_000 == 0) {
-                    buf.append(Integer.toString((nanoValue / 1000_000) + 1000).substring(1));
+                    digits = nanoValue / 1000_000;
```
Comment (bokken) [inline src/java.base/share/classes/java/time/LocalTime.java:1653]: should it be `1_000_000` - breaking on every 3 digits?

**Comment by wenshao** [comment]
Here are the performance numbers running on a MacBook M1 Max, 182.06% performance improvement

```diff
-Benchmark                         Mode  Cnt  Score   Error   Units (master)
-ToStringBench.localTimeToString  thrpt   15  7.752 ? 1.036  ops/ms

+Benchmark                         Mode  Cnt   Score   Error   Units (current ac30276105e541dac48e0d11f4595f93202031b8)
+ToStringBench.localTimeToString  thrpt   15  21.866 ? 0.417  ops/ms +182.06
```

**Review by liach — APPROVED** [review]
The extraction of string length logic to `DecimalDigits` look great.

/reviewers 2 reviewer

**Comment by wenshao** [comment]
Need a second reviewer to approve it. Can anyone take a look at this PR?

**Review by RogerRiggs — APPROVED** [review]
lgtm

**Review by naotoj — APPROVED** [review]

**Pull request closed without merging**